model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/entities/CachedGroup.java

```diff
@@ -21,50 +21,56 @@
 import org.keycloak.models.GroupModel;
 import org.keycloak.models.RealmModel;
 import org.keycloak.models.RoleModel;
+import org.keycloak.models.cache.infinispan.DefaultLazyLoader;
+import org.keycloak.models.cache.infinispan.LazyLoader;
 
-import java.util.HashSet;
+import java.util.Collections;
 import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
 
 /**
  * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
  * @version $Revision: 1 $
  */
 public class CachedGroup extends AbstractRevisioned implements InRealm {
-    private String realm;
-    private String name;
-    private String parentId;
-    private MultivaluedHashMap<String, String> attributes = new MultivaluedHashMap<>();
-    private Set<String> roleMappings = new HashSet<>();
-    private Set<String> subGroups = new HashSet<>();
+
+    private final String realm;
+    private final String name;
+    private final String parentId;
+    private final LazyLoader<GroupModel, MultivaluedHashMap<String, String>> attributes;
+    private final LazyLoader<GroupModel, Set<String>> roleMappings;
+    private final LazyLoader<GroupModel, Set<String>> subGroups;
 
     public CachedGroup(Long revision, RealmModel realm, GroupModel group) {
         super(revision, group.getId());
         this.realm = realm.getId();
         this.name = group.getName();
         this.parentId = group.getParentId();
-
-        this.attributes.putAll(group.getAttributes());
-        for (RoleModel role : group.getRoleMappings()) {
-            roleMappings.add(role.getId());
-        }
-        Set<GroupModel> subGroups1 = group.getSubGroups();
-        if (subGroups1 != null) {
-            for (GroupModel subGroup : subGroups1) {
-                subGroups.add(subGroup.getId());
+        this.attributes = new DefaultLazyLoader<>(source -> new MultivaluedHashMap<>(source.getAttributes()));
+        this.roleMappings = new DefaultLazyLoader<>(source -> source.getRoleMappings().stream().map(RoleModel::getId).collect(Collectors.toSet()));
+        this.subGroups = new DefaultLazyLoader<>(source -> {
+            if (source == null) {
```

Comment:
This applies to the rest of the lazy loaders as well. Maybe worth adding a logic to `DefaultLazyLoader` (a Producer in constructor) for case of `source == null`.